### PR TITLE
Gradient Global Illumination

### LINKED ISSUE #19383: O3de needs an ambient light component to change the color of the pure black formless shadows.
**Is your feature request related to a problem? Please describe.**
The problem is not everyone has RTX for defuse probe grid also at the moment the defuse prob grid is broken. So no global illumination. So all u see in shadows is this pure black and its not ideal for a game. 

**Describe the solution you'd like**
maybe some way to change the ambient black light and change its color and make things hidden by blackness more visible. Or even better some fast cheap global illumination solution.

Image

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by GaianHelmers**
## What does this PR do?

Adds colour based gradient cube map for GI.

## How was this PR tested?

Built it locally and ran it in editor.

https://github.com/user-attachments/assets/6d9f8732-edb2-40c2-b03a-62c7b07c8400

### CODE REVIEW

**Comment by GaianHelmers** [comment]
Realtime Compute Shader driven GI updating. Only dektop compatible, but enables full blown day/night cycle!

https://github.com/user-attachments/assets/dc7aa479-a0b3-4e59-875c-68a4d28d671a

**Comment by GaianHelmers** [comment]
I am having an issue where using a GI component in a surrogate "Editor Only" sky and time system, then entering the play-time, which deletes the editor time one and plants the game-time one, is causing conflicts.

This is the same issue that arises from the Stars Component. They launch , or destroy on top of each other and it messes up the render pipeline: https://github.com/o3de/o3de/issues/19412


https://github.com/user-attachments/assets/9aec1f20-67b4-4fc9-b3a6-e0cc1492ab58


---

I also left the debug logging on. Oops.

**Comment by GaianHelmers** [comment]
Got a fix in.

**Comment by KylerSF** [comment]
Seeing this is more or less an alternative to Global Skylight (IBL) component, this component should also have the **(IBL)** after Gradient GI (IBL) to let people know it is image based lighting GI.

**Comment by KylerSF** [comment]
Transferring my findings while trying to break this feature.

The first issue I almost immediately encountered is a Panic crash. When you are on the mobile non-dynamic update mode and try and scroll the face resolution it works for a small time and then panics and crashes the editor.

The other issue which is not as important as it does not crash anything is when you have both the GradientGI and the IBL global skylight (in the default scene) and you remove the IBL global skylight it also stops rendering the GradientGI unless you change the GradientGI's settings like exposure, or disable and enable the component.
This issue is less important as nobody should have both a Global Skylight IBL and a Gradient GI IBL in their level, if they do on play mode it picks one of them so no UB happens that needs changes right now.

I wanted to point out another bug I found. When you change the face resolution from say 64 to 254 then back to 64, it changes the look, and adds what looks like artifacts from the resizing. Attached pictures of what it looks like before and after the change (You can also make the change happen by scrolling the face resolution scrollbar fast in dynamic update mode)
image
image
This look is only in editor, but also after checking only happens with the Dynamic Update Mode. The Static or mobile safe update mode does not have this issue since it creates a new image every time you update the Face resolution.

The next bug is related to the nodes in script canvas, setting the face resolution is not possible due to a conversion that needs to be made.
image
image
This was only tested with this one node, multiple nodes may need to have this conversion be made internally or setup to only take in an integer value (which is less user friendly in script canvas)


Finally make sure to remove the debugging in the console before removing from draft.

**Comment by KylerSF** [comment]
image

Some thoughts on the naming;
Dropdown (Should not be Static and Dynamic seeing that the static one is just generated, you can dynamically change it during runtime, it is just more expensive than changing the dynamic or compute based one during runtime from the description)
**Static (Mobile-Safe)** > **CPU**
**Dynamic (Desktop GPU)** > **GPU**

Hover Description;
previous

> Static: CPU-generated StreamingImage (mobile-safe, updates on change).
> Dynamic: GPU Compute Pass writing every frame (Desktop only, falls back to Static if unsupported).

This is a great description, but even the Static description states it updates on change, and with testing you could easily update this every frame if you choose to with very little performance impact if any compared to the Dynamic compute pass.
The main difference is the fact that one is CPU based while the other is GPU based.

new

**CPU: CPU-generated StreamingImage (mobile-safe, updates on change)**
**GPU: GPU Compute Pass writing every frame (falls back to CPU if unsupported)**

Very little change to the description just updating it according to the proposed name change. Also is it known that the Dynamic or GPU update mode does not work on mobile devices? For example if it works but may not be the most performant that should be up to the developer, as the Static or CPU update mode already states mobile-safe, Dynamic or GPU update mode should not have to also state that, unless it is fully unsupported. (can a mobile device not use a compute pass?)


The thought of this change is keep it simple and to the point on the dropdown, and when hovering for more details give everything that is needed to understand what both options do. They both can be updated, but the GPU one uses the gpu and updates every frame, a little more expensive and GPU based, but works best when updating this, well every frame.

**Comment by KylerSF** [comment]
Some thoughts on the naming of the component, it probably should end with **(IBL)** like the Global Skylight (IBL) to let developers know it is an IBL solution. This would also convey that they are both using the same underlying technology so you probably should not try and use them together (I have some thoughts on changing some things to level components later, but that is not part of this PR).

**Gradient GI** > **Gradient GI (IBL)**

image

File: Gems/AtomLyIntegration/CommonFeatures/Code/Source/GradientGI/EditorGradientGIComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                        // -- Settings group --
+                        ->ClassElement(AZ::Edit::ClassElements::Group, "Settings")
+                            ->Attribute(AZ::Edit::Attributes::AutoExpand, true)
+                        ->DataElement(AZ::Edit::UIHandlers::Slider, &GradientGIComponentConfig::m_exposure, "Exposure", "IBL exposure in EV stops")
+                            ->Attribute(AZ::Edit::Attributes::ChangeNotify, AZ::Edit::PropertyRefreshLevels::ValuesOnly)
+                            ->Attribute(AZ::Edit::Attributes::SoftMin, -5.0f)
+                            ->Attribute(AZ::Edit::Attributes::SoftMax,  5.0f)
+                            ->Attribute(AZ::Edit::Attributes::Min, -20.0f)
+                            ->Attribute(AZ::Edit::Attributes::Max,  20.0f)
+                        ->DataElement(AZ::Edit::UIHandlers::Slider, &GradientGIComponentConfig::m_faceResolution, "Face Resolution", "Cubemap face size in pixels (4-256)")
```
Comment (KylerSF) [inline Gems/AtomLyIntegration/CommonFeatures/Code/Source/GradientGI/EditorGradientGIComponent.cpp]: Related to my UX comments, "Face Resolution" could just be named "Resolution". It is much easier to connect the word Resolution to the clarity of the GI compared to Face Resolution which people would ask themself what Face? and have to look into the description to understand its usage.
Descriptions should give the full information of what a data element does, while the display name should connect the dots pretty well while also being short.

**Comment by KylerSF** [comment]
Another small nitpick, should the Exposure go low enough to pretty much make it not visible? Right now the cutoff is at -5, but with that on a flat white material mesh, there is still color applied.
image

**Comment by KylerSF** [comment]
I gave this another go today, almost everything seems to be working properly, and from a UI/UX perspective it is much easier to understand all the parameters on this component, but the CPU scrubbing still ends in a crash. The error log reads as;

Exception Description: EXCEPTION_ACCESS_VIOLATION, Attempt to read from address 0x00000240
The memory could not be "read"
Call Stack Trace:
25) 00007FFFF8B816D3 (KERNELBASE) : UnhandledExceptionFilter
24) 00007FFFFBE08C11 (ntdll) : strncpy
23) 00007FFFFBDC1733 (ntdll) : _C_specific_handler
22) 00007FFFFBE063DF (ntdll) : _chkstk
21) 00007FFFFBCB9497 (ntdll) : RtlLocateExtendedFeature
20) 00007FFFFBE05D1E (ntdll) : KiUserExceptionDispatcher
19) 00007FFFD33B39C8 (D3D12Core) : (function-name not available)
18) C:\Users\Kyler\o3de\Gems\Atom\RHI\DX12\Code\Source\RHI\Device.cpp (786) : AZ::DX12::Device::CreateImageReserved
17) C:\Users\Kyler\o3de\Gems\Atom\RHI\DX12\Code\Source\RHI\StreamingImagePool.cpp (485) : AZ::DX12::StreamingImagePool::InitImageInternal
16) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\DeviceResourcePool.cpp (186) : AZ::RHI::DeviceResourcePool::InitResource
15) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\DeviceImagePoolBase.cpp (20) : AZ::RHI::DeviceImagePoolBase::InitImage
14) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\DeviceStreamingImagePool.cpp (95) : AZ::RHI::DeviceStreamingImagePool::InitImage
13) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\StreamingImagePool.cpp (136) : ``AZ::RHI::StreamingImagePool::InitImage'::`2'::<lambda_1>::operator()'::`2'::<lambda_1>::operator()<int,AZStd::intrusive_ptr<AZ::RHI::DeviceStreamingImagePool> >
12) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\StreamingImagePool.cpp (124) : `AZ::RHI::StreamingImagePool::InitImage'::`2'::<lambda_1>::operator()
11) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\ResourcePool.cpp (139) : AZ::RHI::ResourcePool::InitResource
10) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\ImagePoolBase.cpp (20) : AZ::RHI::ImagePoolBase::InitImage
 9) C:\Users\Kyler\o3de\Gems\Atom\RHI\Code\Source\RHI\StreamingImagePool.cpp (117) : AZ::RHI::StreamingImagePool::InitImage
 8) C:\Users\Kyler\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Image\StreamingImage.cpp (148) : AZ::RPI::StreamingImage::Init
 7) C:\Users\Kyler\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Image\StreamingImage.cpp (104) : AZ::RPI::StreamingImage::CreateInternal
 6) C:\Users\Kyler\o3de\Code\Framework\AzCore\AzCore\std\function\function_template.h (172) : AZStd::Internal::function_util::get_invoker<AZStd::intrusive_ptr<AZ::RPI::StreamingImage> __cdecl(AZ::Data::AssetData *),AZStd::Internal::function_util::function_ob
 5) C:\Users\Kyler\o3de\Code\Framework\AtomCore\AtomCore\Instance\InstanceDatabase.h (326) : AZ::Data::InstanceDatabase<AZ::RPI::StreamingImage>::FindOrCreate
 4) C:\Users\Kyler\o3de\Gems\Atom\RPI\Code\Source\RPI.Public\Image\StreamingImage.cpp (31) : AZ::RPI::StreamingImage::FindOrCreate
 3) C:\Users\Kyler\o3de\Gems\Atom\Feature\Common\Code\Source\GradientGI\GradientGIFeatureProcessor.cpp (603) : AZ::Render::GradientGIFeatureProcessor::BuildGradientCubemap
 2) C:\Users\Kyler\o3de\Gems\Atom\Feature\Common\Code\Source\GradientGI\GradientGIFeatureProcessor.cpp (187) : AZ::Render::GradientGIFeatureProcessor::Simulate
 1) C:\Users\Kyler\o3de\Code\Framework\AzCore\AzCore\Jobs\JobFunction.h (43) : AZ::JobFunction<`AZ::RPI::Scene::SimulateJobs'::`6'::<lambda_1> >::Process


Other than the cpu scrubbing issue, from a user perspective this seems ready to go. Though someone with more experience in the c++ department will have to give it a good look.

**Comment by GaianHelmers** [comment]
Ahh yup! I did not do anything with the colour scrubbing. Missed that.

Will definitely tackle that next... It'll be very straightforward, you just buffer the colour changes and only process them into the cubemap every increment. Instead of literally every value change.

**Comment by KylerSF** [comment]
Oh sorry my bad, I misread the bullet points on the discord thread, I thought this wave of fixes included that one. It was a misread on my part.

Everything that was fixed seems to all work properly now though.

**Comment by GaianHelmers** [comment]
> Oh sorry my bad, I misread the bullet points on the discord thread, I thought this wave of fixes included that one. It was a misread on my part.
> 
> Everything that was fixed seems to all work properly now though.

Oh no fault at all, I just missed it. Definitely needs fixing anyways. So glad we can just knock it out and be done with it, haha.

File: Gems/AtomLyIntegration/CommonFeatures/Code/Source/GradientGI/GradientGIComponentController.cpp [hunk 2]
```diff
@@ -12,11 +12,28 @@
...
 #include <Atom/RPI.Public/Scene.h>
 #include <Atom/Utils/Utils.h>
 
+namespace
+{
+    // Normalise a user- or script-supplied asset path to the catalog's product-path convention:
+    // unify separators to forward slashes and strip any leading slashes so the result is
+    // catalog-relative. This lets a manually typed string path and a script-variable path resolve
+    // through the same GetAssetIdByPath lookup regardless of how they were entered.
+    AZStd::string NormalizeAssetPath(AZStd::string_view rawPath)
```
Comment (KylerSF) [inline Gems/AtomLyIntegration/CommonFeatures/Code/Source/GradientGI/GradientGIComponentController.cpp]: Everything looks good from an editor / user standpoint. I could not create any crashes, but I am concerned about this function, what would happen if someone tries to pass in the path /// so firstReal could not find anything what would path.erase() do?

Comment (KylerSF) [reply Gems/AtomLyIntegration/CommonFeatures/Code/Source/GradientGI/GradientGIComponentController.cpp]: I have given this a try, everything seems fine with this function.

Like we talked about in the discord thread, this function and the ibl texture by path may be removed due to it not working properly because by path does not pick up on the asset id. We should get this figured out soon as discussed.

**Comment by GaianHelmers** [comment]
Synched up to dev head. Let's see how builds go. (In case of Qt6 issues.)

Gradient GI systems ready for actual review.

**Comment by GaianHelmers** [comment]
Last commit fixes a regression caused likely by Qt6, where mip generation was causing massive black blocks all over the screen lighting because of misaligned colour checks due to a compression.

Resolution values in CPU mode get snapped to power of twos to fix this issue. Less flexible or intuitive, but gets the job done.

**Comment by KylerSF** [comment]
I want to add on the editor side of things it works properly, scripting also works with no issues. QT6 does cause a few small problems, but it is not due to this addition and just needs some TLC in the Editor department.

I cannot cause any crashes anymore either, so this feature should just needs a code check.

**Comment by nick-l-o3de** [comment]
sorry about needing rebase all the time.

**Comment by nick-l-o3de** [comment]
Is this 100% ready now?

**Comment by GaianHelmers** [comment]
@nick-l-o3de I haven't been able to test this last fix works.

I've put a message in the discord in hopes of getting someone to test to expedite it.
But soon.

**Comment by GaianHelmers** [comment]
Darn. This didn't end up solving the problem. I'll have to try again soon.

**Comment by GaianHelmers** [comment]
After some code review.

### 1 - Refactor/Cleanup
- Removed dead code and relic variables.
- Collapsed redundant methods.
- No longer tries to broker ownership between it and Global Skylight IBL. Was an excessive amount of handling, and dubious at best.

### 2 - Code Review Findings
This one I'll copy paste.
- Confirmed: Reset() didn't stop the feature processor. Render() had no activity gate at all, and OnRenderPipelineChanged re-armed a rebuild, so it resurrected itself with no component alive.
- Differed: property edits do destroy and recreate the component — EditorComponentAdapter::OnConfigurationChanged calls Deactivate()/Activate(). That changed the fix from "destroy the pass on deactivate" to "gate it, and defer teardown".
- Differed: the fallback is black, not the engine grey. Stock ImageBasedLightFeatureProcessor binds null images until something calls Reset(), so a scene with no other IBL provider has no floor.
- Differed: maximized play isn't a PassChanged rebuild — it removes and re-adds the pipeline, then a later rebuild orphans the injected pass with no notification. Instrumentation was needed; three successive event-driven fixes failed.

What made it in.
- Explicit "active but doing nothing" state — Nick's suggested bool, implemented as m_enabled, checked by Simulate, Render, AddRenderPasses and OnRenderPipelineChanged.
- Deactivation genuinely stops output — IBL slots released, compute dispatch silenced, resources freed after a short grace period.
- Deliberately not taken: destroying the pass on deactivate. Because property edits cycle activation, that would rebuild a GPU pass on every slider tick; deferred teardown gets the same result without the churn.

### 3 - Additional Fixes
- Maximized to Editor changing for both GPU and CPU quite stable.
- Deleting Gradient GI clears all injected pass data. Preventing any abandonment of 2 cubemaps worth of data. Left until restart.
- Gradient GI Ownership stable. Can handle any number of concurrent components, and any order of teardown.
- A warning is made when 2 Gradient GI's are present for an extended time. 30 frames. Duplicate GI's is arguably only a mistake, so I wanted to ensure there was reporting on it to prevent silent issues.
- I THINK/HOPE that the Editor-only -> Prefab Spawn -> Editor only edge case is covered. I was unable to replicate it in the GI branch.

---

Video on Discord Thread.

**Comment by nick-l-o3de** [comment]
I'm going to grab and test this again as soon as I can, from a code point of view, it looks good so far.

**Comment by nick-l-o3de** [comment]
I tested this locally and at least on PC DX12, no longer had any issues with game mode enter/leave etc.

**Review by nick-l-o3de — APPROVED** [review]
I can at least approve all of the non-gpu / non rendering parts, they look good to me, and I tested them locally as well.

Would be nice for a graphics reviewer to chime in on the graphics parts, though.
